File: examples/integrators/pattern_stm.rs
//! ```cargo
//! [dependencies]
//! brahe = { path = "../.." }
//! nalgebra = "0.34"
//! ```

//! State Transition Matrix propagation pattern example.
//!
//! Demonstrates the basic pattern for propagating a state transition matrix
//! alongside the state using variational equations.

use brahe::integrators::*;
use brahe::math::jacobian::*;
use nalgebra::{DMatrix, DVector};

fn main() {
    // Dynamics function: Exponential decay dx/dt = -k*x
    let k = 0.1;
    let dynamics = move |_t: f64, state: DVector<f64>, _params: Option<&DVector<f64>>| -> DVector<f64> {
        DVector::from_vec(vec![-k * state[0]])
    };
    // Clone for Jacobian computation
    let dynamics_for_jac = move |_t: f64, state: DVector<f64>, _params: Option<&DVector<f64>>| -> DVector<f64> {
        DVector::from_vec(vec![-k * state[0]])
    };

    // Create Jacobian for variational equations
    let jacobian = DNumericalJacobian::central(Box::new(dynamics_for_jac))
        .with_adaptive(1e-8, 1e-6);

    // Create integrator with Jacobian
    let config = IntegratorConfig::adaptive(1e-12, 1e-11);
    let integrator = DormandPrince54DIntegrator::with_config(
        1,
        Box::new(dynamics),
        Some(Box::new(jacobian)),
        None,
        None,
        config
    );

    // Propagate state and STM
    let t = 0.0;
    let state = DVector::from_vec(vec![1.0]);
    let phi = DMatrix::identity(1, 1); // 1x1 identity matrix
    let dt = 60.0;

    let result = integrator.step_with_varmat(t, state, phi, Some(dt));
    let new_state = result.state;
    let new_phi = result.phi.unwrap();
    let dt_used = result.dt_used;

    println!("Initial state: {:.6}", 1.0);
    println!("State after {:.2}s: {:.6}", dt_used, new_state[0]);
    println!("State transition matrix:");
    println!("  Φ = {:.6}", new_phi[(0, 0)]);
    println!("  (Analytical Φ = {:.6})", (-0.1 * dt_used).exp());
}

File: examples/integrators/pattern_stm.py
#!/usr/bin/env python3
"""
State Transition Matrix propagation pattern example.

Demonstrates the basic pattern for propagating a state transition matrix
alongside the state using variational equations.
"""

import brahe as bh
import numpy as np


def dynamics(t, state):
    """Exponential decay dynamics: dx/dt = -k*x"""
    k = 0.1
    return np.array([-k * state[0]])


# Create Jacobian for variational equations
jacobian = bh.NumericalJacobian.central(dynamics).with_adaptive(
    scale_factor=1e-8, min_value=1e-6
)

# Create integrator with Jacobian
config = bh.IntegratorConfig.adaptive(abs_tol=1e-12, rel_tol=1e-11)
integrator = bh.DP54Integrator(1, dynamics, jacobian=jacobian, config=config)

# Propagate state and STM
t = 0.0
state = np.array([1.0])
phi = np.eye(1)  # Identity matrix
dt = 60.0

new_state, new_phi, dt_used, error_est, dt_next = integrator.step_with_varmat(
    t, state, phi, dt
)

print(f"Initial state: {state[0]:.6f}")
print(f"State after {dt_used:.2f}s: {new_state[0]:.6f}")
print("State transition matrix:")
print(f"  Φ = {new_phi[0, 0]:.6f}")
print(f"  (Analytical Φ = {np.exp(-0.1 * dt_used):.6f})")
